src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/WrapWithTag/RemoteSpanMappingService.cs

```diff
@@ -0,0 +1,138 @@
+    {
+        var generatedDocument = await solution.GetSourceGeneratedDocumentAsync(generatedDocumentId, cancellationToken).ConfigureAwait(false);
+        if (generatedDocument is null)
+        {
+            return [];
+        }
+
+        if (!generatedDocument.Project.TryGetHintNameFromGeneratedDocumentUri(generatedDocument.CreateUri(), out var hintName))
+        {
+            return [];
+        }
+
+        var projectSnapshot = _snapshotManager.GetSnapshot(generatedDocument.Project);
+
+        var razorDocument = await projectSnapshot.TryGetRazorDocumentFromGeneratedHintNameAsync(hintName, cancellationToken).ConfigureAwait(false);
+        if (razorDocument is null)
+        {
+            return [];
+        }
+
+        var documentSnapshot = _snapshotManager.GetSnapshot(razorDocument);
+        var output = await documentSnapshot.GetGeneratedOutputAsync(cancellationToken).ConfigureAwait(false);
+
+        var source = output.Source.Text;
+
+        var csharpDocument = output.GetRequiredCSharpDocument();
+        var filePath = output.Source.FilePath.AssumeNotNull();
+
+        using var results = new PooledArrayBuilder<RazorMappedSpanResult>();
+
+        foreach (var span in spans)
+        {
+            if (RazorEditHelper.TryGetMappedSpans(span, source, csharpDocument, out var linePositionSpan, out var mappedSpan))
+            {
+                results.Add(new(filePath, linePositionSpan, mappedSpan));
+            }
+            else
+            {
+                results.Add(default);
+            }
+        }
+
+        return results.ToImmutableAndClear();
+    }
+
+    public ValueTask<ImmutableArray<RazorMappedEditResult>> MapTextChangesAsync(RazorPinnedSolutionInfoWrapper solutionInfo, DocumentId generatedDocumentId, ImmutableArray<TextChange> changes, CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            solution => MapTextChangesAsync(solution, generatedDocumentId, changes, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<ImmutableArray<RazorMappedEditResult>> MapTextChangesAsync(Solution solution, DocumentId generatedDocumentId, ImmutableArray<TextChange> changes, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var generatedDocument = await solution.GetSourceGeneratedDocumentAsync(generatedDocumentId, cancellationToken).ConfigureAwait(false);
+            if (generatedDocument is null)
+            {
+                return [];
+            }
+
+            if (!generatedDocument.Project.TryGetHintNameFromGeneratedDocumentUri(generatedDocument.CreateUri(), out var hintName))
+            {
+                return [];
+            }
+
+            var projectSnapshot = _snapshotManager.GetSnapshot(generatedDocument.Project);
+
+            var razorDocument = await projectSnapshot.TryGetRazorDocumentFromGeneratedHintNameAsync(hintName, cancellationToken).ConfigureAwait(false);
```

Comment:
Is there any value in extracting out the duplicate code that gets the razor document?

Reply:
Yep, good idea.

Reply:
Actually this can be simplified too, because we can now get the source generated identity from the generated document, rather than round-tripping through the Uri